### ISSUE TITLE
Breaks Jenkins API basic auth

### ISSUE THREAD

**Issue by rossta** [mocleiri/github-oauth-plugin#18]
Before installing the plugin, we were using basic auth with the Jenkins API. After installing the github oauth plugin and updating usernames and tokens as needed, our API requests generate 500 errors (see below). Is this a bug or should we use a different auth strategy for our API requests?

---

GET /api/json (with basic auth headers)

Exception stacktrace: 
java.lang.NullPointerException
at org.jenkinsci.plugins.GithubSecurityRealm.loadGroupByGroupname(GithubSecurityRealm.java:452)
at org.jenkinsci.plugins.GithubSecurityRealm.loadUserByUsername(GithubSecurityRealm.java:421)
at hudson.model.User.impersonate(User.java:251)
at jenkins.security.ApiTokenFilter.doFilter(ApiTokenFilter.java:50)
at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:66)
at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:164)
at winstone.FilterConfiguration.execute(FilterConfiguration.java:194)
at winstone.RequestDispatcher.doFilter(RequestDispatcher.java:366)
at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:81)
at winstone.FilterConfiguration.execute(FilterConfiguration.java:194)
at winstone.RequestDispatcher.doFilter(RequestDispatcher.java:366)
at winstone.RequestDispatcher.forward(RequestDispatcher.java:331)
at winstone.RequestHandlerThread.processRequest(RequestHandlerThread.java:245)
at winstone.RequestHandlerThread.run(RequestHandlerThread.java:148)\n\tat java.lang.Thread.run(Thread.java:679)

**Comment by mocleiri** [comment]
Thanks for you bug report.

I can replicate this error but I don't have an immediate fix.

The ApiTokenFilter (part of Jenkins) is present to load the user based on the token defined in the user screen but it doesn't create a compatibile authentication token for use with the github-oauth-plugin.

The filter creates a User object that is then used to determine if they are authorized to do things.

The null pointer exception is occuring because the GithubSecurityRealm uses the GithubAuthenticationToken to do the user lookups but in the API case it is null.

I will update this ticket when I have a fix for you to test out.

**Comment by mkoppanen** [comment]
I think I am hitting this issue as well. Getting 401 Authorization required with basic auth after enabling the plugin

**Comment by davidjb** [comment]
I'm seeing this same issue when issuing commands via SSH to Jenkins (version 0.11 of the plugin).

Sample traceback:

```
WARNING: SSH command execution failed: null
java.lang.NullPointerException
    at org.jenkinsci.plugins.GithubSecurityRealm.loadGroupByGroupname(GithubSecurityRealm.java:452)
    at org.jenkinsci.plugins.GithubSecurityRealm.loadUserByUsername(GithubSecurityRealm.java:421)
    at hudson.model.User.impersonate(User.java:252)
    at org.jenkinsci.main.modules.sshd.CLICommandAdapter$1.run(CLICommandAdapter.java:30)
    at org.jenkinsci.main.modules.sshd.AsynchronousCommand$1.run(AsynchronousCommand.java:91)
    at java.lang.Thread.run(Thread.java:679)
```

**Comment by djsell** [comment]
I'm running into the same problem.  Any progress or workarounds?  Thanks.

**Comment by blongden** [comment]
Same issue for me too. 401 Authorization required with basic auth.

**Comment by dkador** [comment]
Running into this too.  Would love a fix.

**Comment by zpatten** [comment]
Hitting this as well.

**Comment by rickmzp** [comment]
same here

**Comment by JCook21** [comment]
I have the same issue but with sending commands to Jenkins through Jabber. It's not a deal breaker by a long shot but it would be nice to be able to use this feature.

**Comment by gravityrail** [comment]
Me too. This is a huge problem for us.

**Comment by patcon** [comment]
hm. thought I'd commented here. ditto for me :)

**Comment by mkhattab** [comment]
Crap, I'm having this issue as well. I guess I feel better that I know I'm not the only one and it's related to the Github oAuth plugin.

I'm getting a 401 error.

**Comment by natebeck** [comment]
Awesome work thus far, the plugin is fantastic.  We're running into this issue as well.  We have a script that uses the Jenkins API token to kick off parameterized builds.  Any progress updates?

**Comment by luisfaceira** [comment]
GitHub auth integration is really great, but this making all API access not work is a big turn down.

Would love to volunteer on fixing this, but I've never developed any jenkins plugin! Anyone has any tips on the quickiest way to edit a plugin's code? Do I need to download the source separately, build the project and upload it to jenkins manually each time I make a small change in code?

**Comment by sclasen** [comment]
Any ETA on getting this released?

**Comment by patcon** [comment]
Are you using bundler @sclasen? It should scratch your itch until the next release :)

``` ruby
gem "github-oauth-plugin",
  :git => "https://github.com/mocleiri/github-oauth-plugin.git",
  :ref => "master"
```

**Comment by sclasen** [comment]
thanks!

**Comment by ghost** [comment]
Hey @patcon , how can we upgrade our current plugin with the current master you specified? I'm not sure how to build and install it. Thanks in advance!

**Comment by patcon** [comment]
Errr... @gtzaven it's been awhile since I dropped that in there. I'm not sure what I was thinking at the time... Maybe check out jenkins ruby plugin development? :)

https://github.com/jenkinsci/jenkins.rb/wiki

Ah, ok I think I remember now: Fork the plugin that depends on this, edit it's Gemfile to point to this repo, then build the hpi according to these intstructions:
https://github.com/jenkinsci/jenkins.rb/wiki/Getting-Started-With-Ruby-Plugins

**Comment by ghost** [comment]
Yeah I'm new to this ;) Thanks, that is helpful.

**Comment by charlax** [comment]
@patcon this fix looks great! When will it be released?

**Comment by patcon** [comment]
haha can't take credit (didn't create it it and haven't used it!), and not the maintainer. @mocleiri deserves all the nerd cred and questions :)

**Comment by ghost** [comment]
Since this plugin isn't Ruby, the Ruby Plugins doc is definitely not the way to go :-) Follow these instructions to build the plugin:

https://wiki.jenkins-ci.org/display/JENKINS/Plugin+tutorial#Plugintutorial-BuildingaPlugin

**Comment by patcon** [comment]
haha what can I say -- some people are are fountains of knowledge. I'm apparently a fountain of ignorance :)

**Comment by charlax** [comment]
Thanks to you both!

For the record:
1. Install Maven
2. Run `mvn install` in this repo
3. Watch Maven downloading half of the Internet
4. You now have a `.hpi` file in target
5. Follow the instruction at section _Deploying a custom build of a core plugin_ in https://wiki.jenkins-ci.org/display/JENKINS/Plugin+tutorial#Plugintutorial-BuildingaPlugin

**Comment by dmahlow** [comment]
Any ETA on a release? I'd rather not build and sideload the plugin myself if I can avoid it. :)

**Comment by ibjhb** [comment]
Any ETA on a release? I'd rather not build and sideload the plugin myself if I can avoid it. :)
+1

**Comment by mpsingh01** [comment]
Any ETA on this one? I am also stuck to authenticateusing Jenkins token and getting a HTTP/1.1 500 Internal Server Error.

**Comment by nicolai86** [comment]
+1 on getting this fix out of the door as soon as possible.

**Comment by tinomen** [comment]
+1 please

**Comment by ghost** [comment]
What's the status of this fix? is it supposed to be in the latest plugin release now? It's been causing problems for us for 6 months...

**Comment by TheJefe** [comment]
+1  Any reason for holding up on releasing this?

**Comment by jmgunn87** [comment]
yeah this would be good :godmode:

**Comment by michaeldauria** [comment]
Just so everyone knows, this has been resolved in the official release as of 0.14. This repository is no longer maintained as the README states, head over here: https://github.com/jenkinsci/github-oauth-plugin

**Comment by TheJefe** [comment]
the latest release of the github oauth plugin has been working for me as well.
